File: csharp/Components/Strata.Diagnostic.Core/Evaluation/DatabaseConnectionCheck.cs
//  ##########################################################################
//  # File Name: DatabaseConnectionCheck.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################

using Strata.Diagnostic.Core.Common;
using System;
using System.Data;
using System.Data.SqlClient;

namespace Strata.Diagnostic.Core.Evaluation
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// $comments$
    /// </summary>
    ///  
    public 
    class DatabaseConnectionCheck:
        DiagnosticCheck
    {
        public string ConnectionString { get; set; }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Creates a new <c>DatabaseConnectionCheck</c> instance.
        /// </summary>
        /// 
        public 
        DatabaseConnectionCheck(String name):
            base( name )
        {
            ConnectionString = String.Empty;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Checks if the the specified connection string can successfully
        /// connect to a database.
        /// </summary>
        /// 
        protected override string 
        RunCheck()
        {
		    try
		    {
			   SqlConnection connection = 
                    new SqlConnection(ConnectionString);

                connection.Open();
                
                if ( connection.State == ConnectionState.Open )
                {
                    try
                    {
                        connection.Close();
                    }
                    catch(SqlException) {}

                    return 
                        "Connected using connection string: " + 
                        ConnectionString;
                }
                
                throw 
                    new DiagnosticException( 
                        "Failed to connect using connection string: " + 
                        ConnectionString );

		    }
		    catch (Exception e)
		    {
			    throw new DiagnosticException( "connection failed",e );
		    }        
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Not implemented because <c>DatabaseConnectionCheck</c> 
        /// is not recoverable.
        /// </summary>
        /// 
        /// <exception cref="NotImplementedException">
        /// <c>DatabaseConnectionCheck</c> is not recoverable.
        /// </exception>
        /// 
        protected override string 
        RunRecovery()
        {
            throw new NotImplementedException();
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <c>DatabaseConnectionCheck</c> is not recoverable.
        /// </summary>
        /// 
        /// <returns>False - because check is not recoverable</returns>
        /// 
        protected override bool 
        IsRecoverable()
        {
            return false;
        }
    }
}

//  ##########################################################################

File: java/Components/Strata.Diagnostic.Core/src/strata/diagnostic/core/evaluation/DatabaseConnectionCheck.java
// ##########################################################################
// # File Name:	DatabaseConnectionCheck.java
// #
// # Copyright:	2011, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataInspector Framework.
// #
// #   			The StrataInspector Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataInspector Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataInspector
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata.diagnostic.core.evaluation;

import strata.diagnostic.core.common.DiagnosticException;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;

/**
 * 
 * @author 		
 *     Sapientia Systems
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public 
class DatabaseConnectionCheck 
	extends DiagnosticCheck
{
	private String itsDatabaseUrl;
	private String itsUserName;
	private String itsPassword;
	
	/************************************************************************
	 * Creates a new DatabaseConnectionCheck. 
	 *
	 * @param name
	 */
	public 
	DatabaseConnectionCheck(String name)
	{
		super( name );
		itsDatabaseUrl = "";
		itsUserName    = "";
		itsPassword    = "";
	}

	/************************************************************************
	 *  
	 *
	 * @param dburl
	 */
	public void 
	setDatabaseUrl(String dburl)
	{
		itsDatabaseUrl = dburl;
	}
	
	/************************************************************************
	 *  
	 *
	 * @param userName
	 */
	public void
	setUserName(String userName)
	{
		itsUserName = userName;
	}
	
	/************************************************************************
	 *  
	 *
	 * @param password
	 */
	public void
	setPassword(String password)
	{
		itsPassword = password;
	}
	
	public String
	getDatabaseUrl()
	{
		return itsDatabaseUrl;
	}
	
	public String
	getUserName()
	{
		return itsUserName;
	}
	
	public String
	getPassword()
	{
		return itsPassword;
	}
	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	protected String 
	runCheck() 
		throws DiagnosticException
	{
		@SuppressWarnings("unused")
		Connection connection = null;
		
		try
		{
			connection = DriverManager.getConnection( 
				itsDatabaseUrl,
				itsUserName,
				itsPassword );
		}
		catch (SQLException e)
		{
			throw new DiagnosticException( e );
		}
		
		return "Connected to " + itsDatabaseUrl + " as user " + itsUserName;
	}

	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	protected String 
	runRecovery() 
		throws DiagnosticException
	{
		return null;
	}

	/************************************************************************
	 * {@inheritDoc} 
	 */
	@Override
	protected boolean 
	isRecoverable()
	{
		return false;
	}

}


// ##########################################################################
